### PR TITLE
add option to enable cross thread callbacks

### PR DESCRIPTION
**Pull request by a4z**
This code is marked as experimental

but first experiments are promising, therefore add an option to use that.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -81,14 +81,20 @@ if(DJINNI_WITH_OBJC)
 endif()
 
 # JNI support
-option(DJINNI_WITH_JNI "Include the JNI support code in Djinni support library." off)
+option(DJINNI_WITH_JNI "Include the JNI support code in Djinni support library." OFF)
+
+option (JNI_CPP_THREAD_ATTACH "Add code enableing calling from C++ thread to Java (experimatal)" OFF)
```
Comment (jothepro) [inline CMakeLists.txt:86]:
```suggestion
option (JNI_CPP_THREAD_ATTACH "Add code enableing calling from C++ thread to Java (experimental)" OFF)
```

Comment (a4z) [reply CMakeLists.txt:86]: we have other `off` also :-) 

but we can start here with the to `OFF` transformation :-)

**Comment by jothepro** [comment]
As far as I understand this PR, you are experimenting with calling back into Java from a thread that has been created in C++. Just out of curiosity, can you provide some information on why it is necessary for you?
There seem to be multiple github-issues regarding this topic in the archived djinni repository. It seems like this is not supported by default for good reason. It is recommended to always create new threads in Java, if possible:

- Advice on this topic from @artwyman: https://github.com/dropbox/djinni/issues/176
- Similar discussion: https://github.com/dropbox/djinni/pull/60 (oops I even commented on this one back then... 🤨)

**Comment by a4z** [comment]
Simple answer: people do it (not me directly, personally I avoid this kind of threads like hell) 
But it happens, as we can see who as been involved in discussions on that topic ;-) 

The question for us as library providers is, let user apps crash if they do it, or enable a way to do it with a note like: don't do it.

For dropbox there might have been a clear kind of view: Our project, our rules. 
But now, as a community project,  I think we have to be open for such changes and at least evaluate them.

However, this PR will not be converted back from the _draft_ status until it is clear how we add unit tests into the CI to execute this code. So it solves, as a positive side effect, also the purpose of CI improvement.

**Comment by jothepro** [comment]
I wonder if we should introduce some kind of FAQ page in the docs where common questions are answered, like e.g. *"how do I do multithreading in djinni C++ code? - You DON'T, always start your threads in the host language (Java/Obj-C/C#) if possible!"* Maybe we could stop some people from falling into this pithole (I also fell into it) that way.

**Comment by a4z** [comment]
Yes, this sounds like a good idea. 
Provide it, have a test it compiles when activated, but have an FAQ mentioning to not put yourself in a position where you would need that.

**Review by freitass — APPROVED** [review]
We currently have this "hack" applied to our version of Djinni (still haven't adopted this repo at work). I was planning on fixing the threading before taking the xlcpp version but this PR helps us doing it sooner :) ... Thanks!

File: CMakeLists.txt [hunk 2]
```diff
@@ -101,13 +101,18 @@ endif()
...
-option(DJINNI_WITH_JNI "Include the JNI support code in Djinni support library." off)
+option(DJINNI_WITH_JNI "Include the JNI support code in Djinni support library." OFF)
+
 if(DJINNI_WITH_JNI)
   set_target_properties(djinni_support_lib PROPERTIES
     POSITION_INDEPENDENT_CODE ON)
   target_include_directories(djinni_support_lib PUBLIC "$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/djinni/jni/>")
   target_sources(djinni_support_lib PRIVATE ${SRC_JNI})
   source_group("jni" FILES ${SRC_JNI})
+  option (JNI_CPP_THREAD_ATTACH "Add code enableing calling from C++ thread to Java (experimental)" OFF)
```
Comment (jothepro) [inline CMakeLists.txt:112]:
```suggestion
  option (JNI_CPP_THREAD_ATTACH "Add code enabling calling from C++ thread to Java (experimental)" OFF)
```

**Review by jothepro — APPROVED** [review]
Cool, let's go for it! 
I hope that I'll be able to provide documentation on why this should be avoided soon. It seems to be a very common pithole :/

**Pull request merged**